oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -262,8 +260,118 @@ impl Runtime {
         Ok(())
     }
 
+    /// Returns whether the calling node is allowed to read from the provided channel, according to
+    /// their respective [Labels](oak_abi::label::Label).
+    fn validate_can_read_from_channel(
+        &self,
+        node_id: NodeId,
+        channel_handle: Handle,
+    ) -> Result<(), OakStatus> {
+        debug!(
+            "validating whether node {:?} can read from channel {:?}",
+            node_id, channel_handle
+        );
+
+        // Allow RUNTIME_NODE_ID access to all handles.
+        if node_id == RUNTIME_NODE_ID {
+            return Ok(());
+        }
+
+        let node_infos = self
+            .node_infos
+            .read()
+            .expect("could not acquire lock on node_infos");
+        let node_info = node_infos.get(&node_id).expect("invalid node_id");
+        let node_label = &node_info.label;
```

Comment:
The scope of the read lock on `node_infos` extends to the end of the block, encompassing acquisition of both the `channels.channels` and the `channels.readers` locks.  Maybe encapsulate this into a `get_node_label(node_id)` method that returns a clone?